language/extensions/move-table-extension/src/lib.rs

```diff
@@ -0,0 +1,534 @@
+    let (val, key_size, val_size) = table.contains(table_context, &key)?;
+
+    Ok(NativeResult::ok(
+        table_context
+            .resolver
+            .operation_cost(TableOperation::Contains, key_size, val_size),
+        smallvec![val],
+    ))
+}
+
+fn native_remove(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert!(ty_args.len() == 2);
+    assert!(args.len() == 2);
+
+    let table_context = context.extensions().get::<NativeTableContext>();
+    let mut table_data = table_context.table_data.borrow_mut();
+
+    let key = args.pop_back().unwrap();
+    let handle = get_table_handle(&pop_arg!(args, StructRef))?;
+    let table = table_data.get_or_create_table(context, handle, &ty_args[0], &ty_args[1])?;
+    let (key_size, val_size) = table.remove(table_context, &key)?;
+
+    Ok(NativeResult::ok(
+        table_context
+            .resolver
+            .operation_cost(TableOperation::Remove, key_size, val_size),
+        smallvec![],
+    ))
+}
+
+fn native_destroy(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert!(ty_args.len() == 2);
+    assert!(args.len() == 1);
+
+    let table_context = context.extensions().get::<NativeTableContext>();
+    let mut table_data = table_context.table_data.borrow_mut();
+
+    let handle = get_table_handle(&pop_arg!(args, StructRef))?;
+    let table = table_data.get_or_create_table(context, handle, &ty_args[0], &ty_args[1])?;
+    let (key_size, val_size) = table.destroy(table_context)?;
+    assert!(table_data.removed_tables.insert(handle));
+
+    Ok(NativeResult::ok(
+        table_context
+            .resolver
+            .operation_cost(TableOperation::Destroy, key_size, val_size),
+        smallvec![],
+    ))
+}
+
+// =========================================================================================
+// Helpers
+
+fn get_table_handle(table: &StructRef) -> PartialVMResult<u128> {
+    let field_ref = table
+        .borrow_field(HANDLE_FIELD_INDEX)?
+        .value_as::<Reference>()?;
+    field_ref.read_ref()?.value_as::<u128>()
+}
+
+fn serialize(layout: &MoveTypeLayout, val: &Value) -> PartialVMResult<Vec<u8>> {
+    val.simple_serialize(layout)
+        .ok_or_else(|| partial_extension_error("cannot serialize table key or value"))
+}
+
+fn deserialize(layout: &MoveTypeLayout, bytes: &[u8]) -> PartialVMResult<Value> {
+    Value::simple_deserialize(bytes, layout)
+        .ok_or_else(|| partial_extension_error("cannot deserialize table key or value"))
+}
+
+fn partial_extension_error(msg: impl ToString) -> PartialVMError {
+    PartialVMError::new(StatusCode::VM_EXTENSION_ERROR).with_message(msg.to_string())
+}
+
+fn partial_abort_error(msg: impl ToString) -> PartialVMError {
+    PartialVMError::new(StatusCode::ABORTED).with_message(msg.to_string())
```

Comment:
I believe ABORTED expects a substatus? Not sure if we actually verify it anywhere but IIRC, this status code is for the `Abort` operation, and some spots might freak out (at least in debug mode) without it

Reply:
Done, and yes, the substatus is required.